src/main.cpp

```diff
@@ -641,16 +641,22 @@ unsigned int LimitOrphanTxSize(unsigned int nMaxOrphans) EXCLUSIVE_LOCKS_REQUIRE
 }
 
 
-
-
-
-
-
-bool IsStandardTx(const CTransaction& tx, string& reason)
+bool IsStandardTx(const CTransaction& tx, string& reason, const int nHeight)
 {
-    if (tx.nVersion > CTransaction::MAX_CURRENT_VERSION || tx.nVersion < CTransaction::MIN_CURRENT_VERSION) {
-        reason = "version";
-        return false;
+    bool isOverwinter = NetworkUpgradeActive(nHeight, Params().GetConsensus(), Consensus::UPGRADE_OVERWINTER);
+
+    if (isOverwinter) {
+        // Overwinter consensus rules apply
+        if (tx.nVersion > CTransaction::OVERWINTER_MAX_CURRENT_VERSION || tx.nVersion < CTransaction::OVERWINTER_MIN_CURRENT_VERSION) {
+            reason = "invalid-overwinter-version";
+            return false;
+        }
+    } else {
+        // Sprout consensus rules apply
```

Comment:
`IsStandard` does not currently check if the transaction is `overwintered`, and so it would return `true` for an `overwintered` transaction with version number 2 before Overwinter is activated -- which is invalid as per `ContextualCheckTransaction`.

Is this okay? It doesn't sound like it, unless every place where `IsStandard` applies, `ContextualCheckTransaction` applies also. Example: if this weren't the case in `AcceptToMemoryPool`, we would accept a transaction into the mempool which violates `ContextualCheckTransaction`.  (`ContextualCheckTransaction` does in fact get called by `AcceptToMemoryPool`, but what other places do we call `IsStandard`?)

Reply:
`IsStandard` is a local restriction that is strictly a tightening of the consensus rules (so yes, `ContextualCheckTransaction` should always be called before `IsStandard`), and is only enforced in `AcceptToMemoryPool` (otherwise its rules would be checked on actual blocks, which would make them consensus rules instead of standard rules).